From eae429e18a5ec4b064dd0e74f5147bd7b1e9b44f Mon Sep 17 00:00:00 2001
From: tejohnson <tejohnson@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Tue, 18 Nov 2014 14:20:58 +0000
Subject: [PATCH 357/366] 2014-11-18  Teresa Johnson  <tejohnson@google.com>

	Backport from mainline and gcc-4_9 branch.

	2014-11-13  Teresa Johnson  <tejohnson@google.com>

	PR tree-optimization/63841
	* tree-ssa-strlen.c (strlen_optimize_stmt): Ignore clobbers.

	2014-11-13  Teresa Johnson  <tejohnson@google.com>

	PR tree-optimization/63841
	* testsuite/g++.dg/tree-ssa/pr63841.C: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@217715 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                           |  8 ++++++++
 gcc/testsuite/ChangeLog                 |  8 ++++++++
 gcc/testsuite/g++.dg/tree-ssa/pr63841.C | 35 +++++++++++++++++++++++++++++++++
 gcc/tree-ssa-strlen.c                   |  2 +-
 4 files changed, 52 insertions(+), 1 deletion(-)
 create mode 100644 gcc/testsuite/g++.dg/tree-ssa/pr63841.C

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 3bb275b..3bc25c3 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,11 @@
+2014-11-18  Teresa Johnson  <tejohnson@google.com>
+
+	Backport from mainline and gcc-4_9 branch.
+	2014-11-13  Teresa Johnson  <tejohnson@google.com>
+
+	PR tree-optimization/63841
+	* tree-ssa-strlen.c (strlen_optimize_stmt): Ignore clobbers.
+
 2014-11-16  Eric Botcazou  <ebotcazou@adacore.com>
 
 	* doc/tm.texi.in (TARGET_FLAGS_REGNUM): Move around.
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index a4c8f41..1de0fd4 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,11 @@
+2014-11-18  Teresa Johnson  <tejohnson@google.com>
+
+	Backport from mainline and gcc-4_9 branch.
+	2014-11-13  Teresa Johnson  <tejohnson@google.com>
+
+	PR tree-optimization/63841
+	* g++.dg/tree-ssa/pr63841.C: New test.
+
 2014-11-12  Jakub Jelinek  <jakub@redhat.com>
 
 	PR ipa/63838
diff --git a/gcc/testsuite/g++.dg/tree-ssa/pr63841.C b/gcc/testsuite/g++.dg/tree-ssa/pr63841.C
new file mode 100644
index 0000000..2a2c78f
--- /dev/null
+++ b/gcc/testsuite/g++.dg/tree-ssa/pr63841.C
@@ -0,0 +1,35 @@
+/* { dg-do run } */
+/* { dg-options "-O2" } */
+
+#include <string>
+
+std::string __attribute__ ((noinline)) comp_test_write() {
+  std::string data;
+
+  for (int i = 0; i < 2; ++i) {
+    char b = 1 >> (i * 8);
+    data.append(&b, 1);
+  }
+
+  return data;
+}
+
+std::string __attribute__ ((noinline)) comp_test_write_good() {
+  std::string data;
+
+  char b;
+  for (int i = 0; i < 2; ++i) {
+    b = 1 >> (i * 8);
+    data.append(&b, 1);
+  }
+
+  return data;
+}
+
+int main() {
+  std::string good = comp_test_write_good();
+  std::string bad = comp_test_write();
+
+  if (good != bad)
+    __builtin_abort ();
+}
diff --git a/gcc/tree-ssa-strlen.c b/gcc/tree-ssa-strlen.c
index dbad4a8..fbd51ec 100644
--- a/gcc/tree-ssa-strlen.c
+++ b/gcc/tree-ssa-strlen.c
@@ -1777,7 +1777,7 @@ strlen_optimize_stmt (gimple_stmt_iterator *gsi)
 	    break;
 	  }
     }
-  else if (is_gimple_assign (stmt))
+  else if (is_gimple_assign (stmt) && !gimple_clobber_p (stmt))
     {
       tree lhs = gimple_assign_lhs (stmt);
 
-- 
2.0.3

